### ISSUE TITLE
Add math functions

### ISSUE THREAD

**Issue by faheel** [faheel/BigInt#5]
**Note**: type `T` can be a `BigInt`, `std::string` or ```long long```.

- [x] `pow`
  ```C++
  BigInt pow(const T& base, int exponent);
  ```
  Returns the value of base<sup>exponent</sup>.

- [x] `sqrt`
  ```C++
  BigInt sqrt(const BigInt& num);
  ```
  Returns the square root of `num`.

- [x] `gcd`
  ```C++
  BigInt gcd(const T& num1, const BigInt& num2);
  BigInt gcd(const BigInt& num1, const T& num2);
  ```
  Returns the GCD of `num1` and `num2`.

- [x] `lcm`
  ```C++
  BigInt lcm(const T& num1, const BigInt& num2);
  BigInt lcm(const BigInt& num1, const T& num2);
  ```
  Returns the LCM of `num1` and `num2`.

- [ ] `is_probable_prime`
   ```C++
  bool BigInt::is_probable_prime(size_t certainty);
  ```
  Returns `true` with a certainty of 
![image](https://user-images.githubusercontent.com/11466676/34271444-5845902c-e6b2-11e7-9bac-669b77aa7c08.png)
  for whether this BigInt is prime, based on either the Fermat or Miller-Rabin primality tests. Similar to [Java's `BigInteger.isProbablePrime`](https://docs.oracle.com/javase/7/docs/api/java/math/BigInteger.html#isProbablePrime(int)).

**Comment by ABelmarM** [comment]
Have you started anything?

**Comment by faheel** [comment]
No, I haven't

**Comment by ABelmarM** [comment]
Mind if I start with pow?

**Comment by faheel** [comment]
Whichever way you like. For `pow` use the [binary exponentiation method](https://en.wikipedia.org/wiki/Binary_exponentiation).

**Comment by ABelmarM** [comment]
`pow` function implemented and pull requested.

**Comment by TheWindRider** [comment]
Hello!  Came across this project from [Up For Grabs](http://up-for-grabs.net)

I have some doubt on the new `pow` function. It's doing the `a^n = (-a)^(1/n) if n < 0` transformation, but I think the formula should be `a^n = 1/(a^(-n)) if n < 0`. 

And the test didn't include any `n < 0` cases yet...

**Comment by faheel** [comment]
Yeah, I'm fixing the `pow` function. There are a few more issues with it such as handling 0<sup>0</sup>, 0<sup>-1</sup>, 0<sup>-2</sup> etc.

**Comment by arvindvs** [comment]
Mind if I try to get started on sqrt?

**Comment by faheel** [comment]
@arvindvs Sure, go ahead!

**Comment by faheel** [comment]
Use [Newton's method](https://en.wikipedia.org/wiki/Integer_square_root#Algorithm_using_Newton's_method) for `sqrt()`.

**Comment by arvindvs** [comment]
I'm currently having trouble compiling the code (sorry I'm completely new to open source and contributions ). I have edited the math.test.cpp file and math.hpp file and they are ready to commit to the branch I have created. Is it alright if I go ahead and push this commit without compiling and correcting potential errors, and you can take a look at it?

**Comment by faheel** [comment]
Yes, you can push your changes and create a PR. Then you can comment regarding the issue you're facing in your PR.

**Comment by anandsit043** [comment]
Hi,
I would like to contribute this project. Can I implement gcd  part?

**Comment by faheel** [comment]
@anandsit043 Sure!

**Comment by arvindvs** [comment]
Can I get started on LCM? (I will have to implement GCD in order to make this work. If someone else is working on GCD, I can wait for them to finish. Otherwise I can proceed with both)

**Comment by anandsit043** [comment]
Hi @arvindvs ,
I am working on GCD

**Comment by KingAkeem** [comment]
I would like to take up the is_probable_prime function if no is currently working on it.

**Comment by faheel** [comment]
@KingAkeem Go for it, as it doesn't seem like anyone else is working on it.

**Comment by KingAkeem** [comment]
Awesome, I'm working on implementing Rabin-Miller primality testing now.

**Comment by KingAkeem** [comment]
Should is_probable_prime be declared as a public member function but defined inside of math.hpp?

**Comment by anandsit043** [comment]
@faheel  @arvindvs 
I have implemented gcd, how do I push these changes, I am getting permission denied error.

**Comment by faheel** [comment]
@KingAkeem Yes.

**Comment by anandsit043** [comment]
Hi @arvindvs, 
Should I implement the lcm part or you are done with it?

**Comment by arvindvs** [comment]
@anandsit043 Sorry for the late response. Go ahead, I have not yet completed LCM.

**Comment by AanjaneyaSinghDhoni** [comment]
how can i contribute to this project. I new to open source. Please guide.

**Comment by faheel** [comment]
@AanjaneyaSinghDhoni All of the functions mentioned in this issue have been implemented (PR for `is_probable_prime` is under review).

If you'd like to contribute, you can try implementing the functions mentioned in issues #18 and #19. If you need some more info regarding implementation details, comment on the respective issues.

**Comment by cedrikaagaard** [comment]
Hello, `is_probable_prime ` is not yet implemented right? Would it be okay if I gave it a shot?

**Comment by ankur54** [comment]
Hello! Is this issue still open? I would like to contribute to it.

**Comment by riddhyasona** [comment]
Hello @faheel sir, I am new to contributions and cmake, though I have a very good experience in coding with c++. I would like to take up the task of building a function out of the list mentioned above or if you can suggest any new function. Or if there's any other issue that might suit me, please suggest.

**Comment by liamob** [comment]
Hello! I'll give the "is_probable_prime" function a try! I'll send a message when it's finished.

**Comment by githubjsl118** [comment]
Hello @faheel I have implemented and tested the is_probable_prime function, however it is difficult to test using a very large number due to the incredibly significant runtime caused when you take the   (randomNumber) to the power of another amount.   

On the wikipedia page (https://en.wikipedia.org/wiki/Miller–Rabin_primality_test) it would be where the pseudocode says x ← ad mod n
Any tips?

**Comment by youcefl** [comment]
> 
> 
> Hello @faheel I have implemented and tested the is_probable_prime function, however it is difficult to test using a very large number due to the incredibly significant runtime caused when you take the (randomNumber) to the power of another amount.
> 
> On the wikipedia page ([https://en.wikipedia.org/wiki/Miller–Rabin_primality_test](https://en.wikipedia.org/wiki/Miller%E2%80%93Rabin_primality_test)) it would be where the pseudocode says x ← a^d mod n
> Any tips?

You have to implement and use a modular exponentiation i.e. when computing X = A^B mod N, you do not compute A^B and then take the remainder modulo N, you loop on the binary representation of B squaring and/or multiplying by A mod N depending on whether the current bit is set.
Have a look at https://en.wikipedia.org/wiki/Modular_exponentiation and/or https://www.geeksforgeeks.org/modular-exponentiation-power-in-modular-arithmetic/.
